### PR TITLE
fix(forms): unsubscribe from async validator when the directive is destroyed

### PR DESCRIPTION
**Pull request by arturovt**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
- [x] Bugfix

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/forms/src/directives/shared.ts [hunk 1]
```diff
@@ -255,6 +255,7 @@ function setUpModelChangePipeline(control: FormControl, dir: NgControl): void {
   // Register a callback function to cleanup onChange handler
   // from a control instance when a directive is destroyed.
   dir._registerOnDestroy(() => {
+    control._cancelExistingSubscription();
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/shared.ts:258]: @arturovt thanks for creating this PR! Please find my reply below.

I think that might be a problem for cases when one control (for ex. `FormControl` instance) is used in multiple places in a template, something like:

```
// in TS:
@Component({ ... })
class Comp {
  control = new FormControl('value');
}

// in template:
<input *ngIf="conditionA" [formControl]="control" id="input-a" viewValidator>
<input *ngIf="conditionB" [formControl]="control" id="input-b" viewValidator>
```

Let's imagine both `conditionA` and `conditionB` were `true` originally and `conditionB` became `false` at some point. This would cause async validator subscription to be cancelled, thus loosing the state update for the control instance (which is still needed for the `input-a`).

Could you please try to create a test with this scenario?

Also, cancelling the subscription would cause control instance state to be stale, so even if an instance is used only once and is no longer active, it might be used as a directive input later (for ex. FromControlDirective).

I'm also a bit confused how the leak happens now and why extra stuff is retained in memory. @arturovt could you please elaborate on this more?

Thank you.

Comment (arturovt) [reply packages/forms/src/directives/shared.ts:258]: Hey @AndrewKushnir 

I'll give you a small and dummy example:

image

Unfortunately, we do have a lot of code like this (but much more complex that is dangerous to refactor) in a system and we can't get rid of it, some kind of legacy code.

I think that this would an expected behavior to unsubscribe from the observable when the directive gets destroyed.

Comment (arturovt) [reply packages/forms/src/directives/shared.ts:258]: I tried your example and of couse it behaves like you said :smile: 

I see that _on change_ functions are stored in array. Do you think it would be a good way to store subscriptions in some appropriate data structure? So I guess there should be 2 subscriptions in your example for 2 different form controls, when the first one is destroyed then it should unsubscribe from the `_asyncValidationSubscription` that is related to the 1st form control.

Comment (AndrewKushnir) [reply packages/forms/src/directives/shared.ts:258]: Hi @arturovt, thanks for additional info.

I think with the current internal structure it might be tricky to solve this issue for all cases, but we can still make an improvement for a large portion of use-cases.

Let's imagine we have the following setup (similar to the one I mentioned in my previous comment, but now the `FormControl` is created with an additional async validator):

```
// in TS:
@Component({ ... })
class Comp {
  control = new FormControl('value', null /* sync validator */, ownAsyncValidator);
}

// in template:
<input *ngIf="conditionA" [formControl]="control" id="input-a" viewAsyncValidatorA>
<input *ngIf="conditionB" [formControl]="control" id="input-b" viewAsyncValidatorB>
```

Assuming `conditionA` and `conditionB` were `true` originally, the `FormControl` instance would have all async validators attached (i.e. `[ownAsyncValidator, viewAsyncValidatorA, viewAsyncValidatorB]`). The current logic [combines all validators into a single validator fn](https://github.com/angular/angular/blob/master/packages/forms/src/model.ts#L449), which gets executed and [there is 1 subscription](https://github.com/angular/angular/blob/master/packages/forms/src/model.ts#L771-L778) which handles the result.

Now if `conditionA` turns `false` while previous async validators are in progress (i.e. the subscription exists), ideally we should:
- remove `viewAsyncValidatorA` from the list of validators attached to the `FormControl` instance (this cleanup happens already)
- if the `viewAsyncValidatorA` validator was still "in progress", we should cancel the request, but **keep** the remaining ones. This is a bit problematic right now due to the way we combine validator functions. Changing this would be quite tricky: we may need to do that in stages where we first turn existing functionality (without introducing unsubscribe actions) to more granular validator-based subscriptions and after that add corresponding cleanup step.

However the good news is that we can still do the cleanup when we detect that the subscription exists, but no async validators are currently present (which means that they were removed prior), which would be the case when async validators were added by a directive that is being destroyed. This should still cover a lot of cases (probably the most common ones), so the code may look like this:

```
  /**
    * TODO: describe what "force" argument is used for.
    * @internal
    */
  _cancelExistingSubscription(force: boolean): void {
    if (this._asyncValidationSubscription && !(force && this._rawAsyncValidators.length > 0)) {
      this._asyncValidationSubscription.unsubscribe();
      this._hasOwnPendingAsyncValidator = false;
    }
  }
```
The `force` flag should be `false` when calling from within the `dir._registerOnDestroy` callback (or may be we can move the call to `cleanUpValidators`?) and `true` in other cases.

Note: I've used `FormControl` in my comment, but the same is applicable to `FormGroup` and `FormArray` classes as well (it'd be great to add tests for other classes as well).

Please let me know what you think.

Thank you.

Comment (arturovt) [reply packages/forms/src/directives/shared.ts:258]: I do like this approach since it won't bring any breaking change and as you said the common case (where there is 1 `formControl` binding) will be solved.

I will test all classes locally and will push my changes to the branch so we'll be able to discuss it.

Cheers!

Comment (arturovt) [reply packages/forms/src/directives/shared.ts]: Hey Andrew @AndrewKushnir . I've pushed new changes. Could you have a look please when you're free.

File: packages/forms/test/reactive_integration_spec.ts [hunk 2]
```diff
@@ -4358,6 +4358,369 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
...
+          validate() {
+            return new Observable<null>(() => {
+              // Note that we intentionally do not call `observer.next` and `observer.complete`
+              // so the subscription will be in non-closed status.
+              return () => {
+                this.unsubscribeCalledTimes++;
+              };
+            });
+          }
+        }
```
Comment (AndrewKushnir) [inline packages/forms/test/reactive_integration_spec.ts:4383]: I think it'd be a bit more easier to work with tests if we log validation start and cancel operations (and cleanup logger during and after each test as needed):
```suggestion
        const logger: string[] = [];
        // ...
        class AppAsyncValidator implements AsyncValidator {
          validate() {
            logger.push('validation started');
            return new Observable<null>(() => {
              // Note that we intentionally do not call `observer.next` and `observer.complete`
              // so the subscription will be in non-closed status.
              return () => {
                logger.push('validation cancelled');
              };
            });
          }
        }
```

File: packages/forms/test/reactive_integration_spec.ts [hunk 3]
```diff
@@ -4358,6 +4358,369 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
...
+               const appAsyncValidator = div.injector.get(AppAsyncValidator);
+
+               fixture.componentInstance.visible = false;
+               fixture.detectChanges();
+
+               // It was called forcely from the `updateValueAndValidity()` method.
+               expect(ownAsyncValidatorUnsubscribeCalledTimes).toEqual(1);
+
+               // It doesn't unsubscribe from the observable since FormGroup has its
+               // own validator function.
```
Comment (AndrewKushnir) [inline packages/forms/test/reactive_integration_spec.ts:4495]: Note: I've created a new ticket https://github.com/angular/angular/issues/40943 to capture the proposal to make subscriptions more granular. You can refer to that ticket in this comment (and in other comments in your PR) to provide additional context, for example:
```suggestion
               // It doesn't unsubscribe from the observable since FormGroup has its
               // own validator function (which gets combined into a single validator function in AbstractControl),
               // so we can not unsubscribe without stopping own async validator.
               // See https://github.com/angular/angular/issues/40943 for additional info.
```

Comment (AndrewKushnir) [reply packages/forms/test/reactive_integration_spec.ts:4495]: I also think that it'd be great to split tests into 2 sub-groups: where we verify that subscription is removed and where we keep subscription due to internal data structures, something like this:
```
  describe('complete cleanup of async validation subscription', () => { ... });

  // Tests below should be revisited as a part of https://github.com/angular/angular/issues/40943,
  // currently we just verify that we are not cancelling async validation altogether if there are own control validators.
  describe('partial cleanup (async validation subscription retained)', () => { ... });
```

**Review by AndrewKushnir — COMMENTED** [review]
@arturovt thanks for additional updates. The code looks good, thanks for adding types on internal fields, extending tests coverage and writing docs 👍 

I've left a few comments, please have a look when you get a chance.

Note: I've also created https://github.com/angular/angular/issues/40943 to capture additional changes that should be made in the future to achieve more granular subscribe/unsubscribe actions.

**Comment by arturovt** [comment]
@AndrewKushnir I've updated the PR per your review. Could you have a look please.

**Comment by AndrewKushnir** [comment]
@arturovt thanks for additional changes. The main code looks good, so I've started tests in [Google's codebase](http://test/OCL:358951769:BASE:358951780:1614048762251:47022269) (internal-only link) to see if there are any patterns we may need to take into account. I will have a look at the tests within the next day or so and provide some additional feedback. Thank you.

**Comment by AndrewKushnir** [comment]
@arturovt quick update: running tests in Google's codebase caused a few failures that I'm investigating, but it will take some time (a couple days). Will keep you updated. Thank you.

**Comment by AndrewKushnir** [comment]
@arturovt I performed quick investigation and found the origin of the problem (and actually we can reproduce it with our unit tests too). The problem is that when we start async validator, we set the status of a control to `PENDING` here:
https://github.com/angular/angular/blob/011a527497aa9247fa6883a3ce5dbad5e297bc5b/packages/forms/src/model.ts#L767-L769
Later when we cancel the subscription as a result of a directive destroy (and there are no other validators), we stop the subscription, but the status still remains `PENDING`. You can reproduce this by adding a line into one of the unit tests to check control status after subscription cancellation scenario.

I believe the expected behavior in this case is to restore the previous status of a control, but it might get quite tricky in case a new async validation is started while the previous one is running (for ex. by calling `updateValueAndValidity` synchronously multiple times), so it'd be very error-prone to maintain the right status.

After looking at the code more, it feels like we would not be able to land this change without additional architecture changes and if that's the case, we can revisit this as a part of #40943.

@arturovt what do you think?

**Comment by AndrewKushnir** [comment]
Sorry for spamming, just one more thought :)

I believe this memory leak issue could actually be resolved in the directive instance with `ngOnDestroy`, for example (using an async validator from the tests):

```
        @Directive({
          selector: '[app-async-validator]',
          providers: [{
            provide: NG_ASYNC_VALIDATORS,
            multi: true,
            useExisting: forwardRef(() => AppAsyncValidator)
          }]
        })
        class AppAsyncValidator implements AsyncValidator {
          validate() { /* init async validation (HTTP request, establish WebSocket connection) */ }
          ngOnDestroy() { /* close open connections, do the necessary cleanup */}
        }
```

This looks reasonable to me since developers have full control over validation details and can perform the necessary cleanup. From the Forms perspective, the subscription will still exist after directive destroy (thus retaining some memory), but it would not cause issues like open WebSocket connections. We'd still need to look at the best ways to do cleanup in Forms, it just feels like this memory leak is not severe.

Would that help in your real app?

**Comment by arturovt** [comment]
Hi Andrew. Oh, sorry to hear that tests have failed. Do you think we could at least update the documentation to mention that developers have to unsubscribe from streams inside async validators?
```ts
@Directive({
  selector: '[app-async-validator]',
  providers: [
    {
      provide: NG_ASYNC_VALIDATORS,
      multi: true,
      useExisting: forwardRef(() => AppAsyncValidator),
    },
  ],
})
class AppAsyncValidator implements AsyncValidator {
  private destroy$ = new Subject();

  validate() {
    return this.establishSomeConnection().pipe(
      switchMap(() => validateDataAndReturnValidationErrorsOrNull()),
      takeUntil(this.destroy$)
    );
  }

  ngOnDestroy() {
    this.destroy$.next();
  }
}
```
They might not (or don't) know that control doesn't unsubscribe from async validation when the directive gets destroyed.

**Comment by AndrewKushnir** [comment]
@arturovt agree, I think updating docs is a great idea 👍 

I did a quick search and it looks like we have some related info in the ["Creating Asynchronous validators" section](https://angular.io/guide/form-validation#creating-asynchronous-validators):

```
The observable returned must be finite, meaning it must complete at some point. To convert an infinite observable into a finite one, pipe the observable through a filtering operator such as first, last, take, or takeUntil.
```

This is an important piece of information that we can add to the `AsyncValidator` and `AsyncValidatorFn` docs (we can rephrase it as needed):

- https://angular.io/api/forms/AsyncValidator
- https://angular.io/api/forms/AsyncValidatorFn

The use-case that you mentioned originally (validation through persistent WebSocket connection) doesn't match that note which says that the observable should be finite, so I don't think we can add extra recommendations to the docs on how to handle persistent connections (in a context of validators).

I think it's still ok to perform validation through WebSockets, it's just a use-case where you'd need to have more handling on validator's side to manage connection state, since current Forms logic is designed to work with finite observables.

So to sum up:

- Let's update `AsyncValidator` and `AsyncValidatorFn` docs with the note that observables should be finite
- I think that issue #40943 is still relevant and we can look into that later (since it needs design work and will most likely introduce a breaking change)
- I will close this PR since we won't be able to land it at this time

@arturovt thanks again for creating this PR! Even though we ended up not merging this change, this PR was still very useful as we've identified potential areas of improvement.
